### ISSUE TITLE
The #369 units error has a third syntactic position and no guard covers it: 36 call-site scalars, 2 in teaching notebooks

### ISSUE THREAD

**Issue by suchethac** [suchethac/tengri#1924]
## The gap

#1819 fixed 125 sites where the #369 rename (`sfh_*_log_peak_sfr` → `log_total_mass`,
`log10(SFR/Msun/yr)` → `log10(M*/Msun)`) carried the range across unconverted, and
`tools/check_param_ranges.py` now guards that shape — a **prior handed to a call site**.
`tools/check_param_defaults.py` guards the same rule one step earlier, at a **signature
default**.

The same number gets written in a third place that neither guard can see: a **bare
scalar at a call site** — a kwarg, a dict literal, or a subscript assignment. There is
no `Uniform(...)` and no signature involved, so both guards walk past it.

## Verified by fault injection, not inferred

Staged a probe under `scripts/` (both guards discover files via `git ls-files`, so it
had to be tracked to be scanned):

```python
def positive_control():
    return dict(sfh_tsnorm_log_total_mass=Uniform(-1.0, 2.5))   # the #1819 shape

def gap_kwarg():
    return dict(sfh_tsnorm_log_total_mass=1.2)

def gap_dict_literal():
    return {"sfh_tsnorm_log_total_mass": 1.2}

def gap_subscript_assign(params):
    params["sfh_tsnorm_log_total_mass"] = 1.2                   # the notebook shape
    return params
```

With the positive control present, `check_param_ranges.py` goes **red** and names the
file:

```
check_param_ranges: 1 call-site prior(s) disjoint from the declared prior
  scripts/_probe_param_gap.py:13
      sfh_tsnorm_log_total_mass = [-1, 2.5]   declared support [7, 12.5]
```

Delete that one function and leave the three gap shapes — same file, same parameter,
same units error, a value six orders of magnitude outside the declared support:

```
check_param_ranges:      OK -- 2518 call-site prior(s) in 2640 files all overlap ...
check_param_defaults:    OK — every signature default is inside its declared prior.
check_param_free_priors: OK: 361 parameters, 78 freeable by 'all_params: FREE'.
```

Green here means blind, not clean. The probe was removed; nothing is committed.

## The 36

Discriminator: a file that **declares** a mass-scale prior (`Uniform(lo, hi)` with
`lo >= 6`) and **also** assigns a scalar outside it. A file contradicting its own
declaration cannot be explained by a differing convention.

| count | file |
|---|---|
| 13 | `tests/integration/test_components_vs_legacy.py` |
| 10 | `analysis/profile_all_components.py` |
| 3 | `tests/crossval/test_bagpipes_feature_parity.py` |
| 2 | `scripts/isolate_nan_sfh.py` |
| 1 | `notebooks/tutorials/notebook_code/01_quickstart.py` |
| 1 | `notebooks/tutorials/notebook_code/04_the_forward_model.py` |
| 1 | `scripts/diagnose_jit_cache_hits.py` |
| 1 | `tests/components/sps/test_stellar_precomp_bounds.py` |
| 1 | `tests/contract/test_stellar_precomp_contract.py` |
| 1 | `tests/integration/test_convention_end_to_end.py` |
| 1 | `tests/integration/test_sed_model_build.py` |
| 1 | `tests/physics/conservation/test_stellar_mass_balance.py` |

`src/` is clean, same as #1819.

## The two that actually teach a wrong number

```python
# notebooks/tutorials/notebook_code/04_the_forward_model.py:145
# Override tsnorm to a typical star-forming galaxy (still forming stars now)
params["sfh_tsnorm_log_total_mass"] = jnp.array(1.2)
```

`log10(M*) = 1.2` is **16 Msun**, described in the comment as a typical star-forming
galaxy, and outside the `Uniform(7.0, 12.5)` this same file declares ~20 lines above.
`notebooks/tutorials/notebook_code/01_quickstart.py:462` is the same value in the same
shape. `1.2` is a plausible `log10(SFR)`, which is the #369 signature exactly.

These two are the ones a reader copies.

## What this issue does not claim

The other 34 are **not** asserted to be bugs. Most are in shape-comparison tests and
profiling code where the absolute mass scale cancels, and CLAUDE.md explicitly sanctions
a unit-mass normalization (`Fixed(0.0)` on a `log_total_mass` for crossval). They need
per-site judgement, not a sweep.

A first, cruder version of this scan returned **338** hits and was worthless — it parsed
`"sfh_tsnorm_log_total_mass": 1.0 + 10.0` (actual value 11.0) as `1.0`. Any fix should
re-derive the list rather than trust a number in prose.

## Suggested shape of the fix

Two separable pieces:

1. **The two notebooks** — a real teaching defect, fixable now, independent of the rest.
2. **The guard gap** — the interesting half. A fourth guard bolted on would work, but
   `check_param_defaults.py`, `check_param_ranges.py` and this would then be three copies
   of one rule ("a number written beside a parameter must be a value that parameter
   admits") differing only in syntactic position. The same pattern already bit
   `tools/check_dust_law_kwargs.py`, which goes red on `law_fn(wave, n_slope=...)` and
   green on the identical defect spelled `law_fn(**dust_kw)`. Worth considering whether
   the position should be the variable rather than the guard.

A guard on position 3 must be quiet enough to leave on: the unit-mass-normalization
convention is legitimate and common, so the rule likely has to be the file-internal
contradiction used above (declares a mass-scale prior *and* contradicts it), not a bare
range check.

**Comment by suchethac** [comment]
**Correcting my own framing.** This issue said the call-site scalar position is one "no guard covers", implying an oversight. It is not — `tools/check_param_ranges.py` addresses it explicitly under *"Why pinned scalars are out of scope"*, and the reasoning is sound:

> A range says *fit this parameter somewhere in here*. One sharing no point with the declaration is incoherent on its own terms. A pinned scalar says *this parameter is not being fit at all, and I want exactly this value* — and choosing a value outside the declared range is a normal thing to do.

It gives three legitimate categories (unit-mass normalization for crossval, CIGALE reference offsets, tests asserting an override beats the declared prior), measures **71** such scalars in the tree, and concludes: *"a guard whose allowlist rivals its findings teaches people to add entries rather than read them."*

That reasoning extends to most of my 36. `analysis/profile_all_components.py` calling `dpl(age_grid, log_total_mass=1.0)` for profiling is the same category — the absolute scale is irrelevant to what the call measures. My scan's discriminator (file-internal contradiction) does not distinguish those from real defects, so **the 36 figure should not be read as 36 defects.** I flagged that risk in the original text; the guard docstring makes it concrete.

**What survives, and it is much narrower.** The defective pattern is not "a scalar outside its declared support". It is *a mock **truth** placed outside the prior it will then be fit under* — unrecoverable by construction, so the fit cannot do anything but fail or mislead. That is a different claim from the one this issue makes, and it applied to exactly two sites:

```
notebooks/tutorials/notebook_code/01_quickstart.py:462   true_params_stoch[...] = jnp.array(1.2)
notebooks/tutorials/notebook_code/04_the_forward_model.py:145  params[...] = jnp.array(1.2)
```

Both generate a mock and then fit it under `Uniform(7.0, 12.5)` declared in the same file, and both are commented *"a typical star-forming galaxy"* while setting 16 M☉.

**Fixed** on `fix/param-scalar-position`: both now use `10.5`, which is not a guess — `01_quickstart.py:151` already used `jnp.array(10.5)` for the parametric mock in the same notebook, so the file supplied its own convention and line 462 was the outlier.

**I am not adding a guard for this.** Detecting "truth outside the prior it is fit under" needs dataflow (value → mock → fit), and a static rule broad enough to catch it re-flags the 71 sites `check_param_ranges` deliberately left alone. Repeating a rejected design with a fresh allowlist would be worse than the gap.

**One real gap did turn up while fixing this**, unrelated to parameters — see the follow-up comment.

**Comment by suchethac** [comment]
**The real gap found while fixing this: `notebooks/tutorials/` has no mirror↔notebook sync guard, and drifted for months.**

Fixing the two values meant syncing the paired notebooks. `jupytext --sync` then propagated far more than my change — the whole `Fitter` → `ForwardModel` migration from `67550696e` (#1341) had never reached the `.ipynb`:

```diff
-    "    Fitter,"                                  +    "    ForwardModel,"
-    "fitter_param = Fitter(model_param, ...)"      +    "forward_param = ForwardModel.build(sed=model_param)"
-    "result_map_param = fitter_param.run(\"map\")"  +    "result_map_param = forward_param.fit(..., method=\"map\")"
-    "fitter_param.compile(verbose=False)"          +    "forward_param.prewarm(method=\"vi\")"
```

So `notebooks/tutorials/01_quickstart.ipynb` has been teaching a superseded API since #1341, while its own mirror taught the current one. The same commit is why cell 7 still held `jnp.array(1.2)` when the mirror had been corrected to `10.5` — I found that drift only because the two disagreed.

**Why nothing caught it.** `tools/check_notebook_renders.py` pairs `notebooks/*.py` against `docs/spine/**/*.ipynb` via `SPINE_SLUGS`. `notebooks/tutorials/` is outside that domain entirely — so the pairing declared in the notebook's own metadata (`"formats": "notebook_code//py:percent,ipynb"`) is enforced by nothing. This is the guard-domain trap: the check is real, passes, and simply never looks here.

**Caveat on the fix.** Cell *sources* now match their mirrors; the *outputs* were produced by the old `Fitter` path and are stale. They were already stale in a different way (source and outputs agreed with each other but not with the mirror). These notebooks are unpublished staging — outside `docs/`, absent from any toctree — so no published page is affected, but they would need a re-run to be internally consistent. I have not re-run them: `01_quickstart` performs MAP + geoVI + NUTS fits, and CLAUDE.md warns of 20+ GB NUTS warmup peaks.

Worth a small guard extending the render check to any `.ipynb` whose jupytext metadata declares a pairing, rather than a hardcoded slug list — that rule would have caught this in 2026-05 and needs no allowlist. Filing separately rather than widening this issue.

**Comment by suchethac** [comment]
Claiming the sync-guard gap (gallery-overhaul session): a mirror↔notebook sync guard for notebooks/tutorials/ so this drift class (the Fitter→ForwardModel migration never reaching the .ipynb) goes red in CI instead of surviving for months.

**Comment by suchethac** [comment]
Closing — all three threads this issue ended up carrying are resolved (verified on current main, 2026-08-21):

1. **The two real defective values** (tutorial mocks outside the prior they are fit with) were fixed in #1941 (merged 2026-08-18).
2. **The original premise — that the bare-scalar call-site position is an uncovered guard gap — was self-corrected in this thread**: `tools/check_param_ranges.py` excludes pinned scalars deliberately and documents why (unit-mass normalizations, reference offsets, override tests; 71 legitimate cases measured vs a handful of candidate defects), and that reasoning was accepted here as extending to most of the 36 flagged sites.
3. **The real gap this issue discovered** — notebooks/tutorials/ had no mirror↔notebook sync guard and drifted for months (a superseded API taught since #1341) — shipped in #1945 (merged 2026-08-18): `tools/check_notebook_pairing.py`, CI-wired, with the three drifted tutorials repaired; independently re-verified passing on current main (27 pairings hold).
